### PR TITLE
[WIP] Morgbranch

### PR DESCRIPTION
**Pull request by mlp6**
starting to create some python functions

### CODE REVIEW

File: phantom2.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Separate Phantom2 Class script, will eventually be integrated with phantom.py
+
+import numpy as np
+from scipy.interpolate import RegularGridInterpolator
+
+
+def my_interpn(x, y, z, data, pts):
+    # Input x, y, z - all np.ndarray vectors of xyz coordinates
```
Comment (mlp6) [inline phantom2.py]: convert function comments to Sphinx docstring

File: test_phantom2.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    z = np.linspace(-2, 2, 5)
+
+    data = f3(*np.meshgrid(x, y, z, indexing='ij', sparse=True))
+    pts = np.array([[1, 0, 0], [0.5, 0.5, 0.5], [-1.5, 0.3, 1.7]])
+
+    # print(my_interpn(x, y, z, data, pts))
+    assert my_interpn(x, y, z, data, pts)[0] == 1
+    assert my_interpn(x, y, z, data, pts)[1] == 1.5
+    assert my_interpn(x, y, z, data, pts)[2] == 0.5
+    print("Test3 pass")
```
Comment (mlp6) [inline test_phantom2.py]: @mjr52 Do not need to print explicit passing.

File: test_phantom2.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    # Ones matrix data
+    return 0*x + 0*y + 0*z + 1
+
+
+def f3(x, y, z):
+    # Linear data
+    return x + y + z
+
+
+def test_1():
```
Comment (mlp6) [inline test_phantom2.py]: @mjr52 Give tests more descriptive names.

**Review by mlp6 — COMMENTED** [review]
Great first pass in python!

File: tests/test_phantom.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+nodoutfile = '%s/nodout2' % myPath
+
+# Make disp.dat.xz file
+from fem.post.create_disp_dat import create_dat, parse_line
+create_dat(dispout="disp.dat.xz", nodout="nodout")
+
+# Make instance of Phantom class
+p = Phantom(scat_density=10,nodesdynfile=nodefile, dispdatfile = dispfile, delta_xyz = (10, 5, 100), nodout= nodoutfile)
+
+def test_mytest():
```
Comment (mlp6) [inline tests/test_phantom.py]: This does nothing.

File: phantom.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
+import numpy as np
+import h5py
+from scipy.interpolate import RegularGridInterpolator
```
Comment (mlp6) [inline phantom.py]: Instead of importing these "globally", I would import them "as needed" in individual methods of the class.

File: phantom.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        self.n_scats = int(np.round(tracking_volume*self.scat_density , decimals = 6))
+
+    def create_scatterers(self):
+        """create scatterrers within phantom_bounds using explicitly-seeded RNG
+        Using the explicit RNG seed ensures identical scatterer location in subsequent runs if needed.
+
+        """
+        np.random.seed(self.rng_seed)
+        scatterers = np.random.random((self.n_scats, 3))
+        lims = self.lims
```
Comment (mlp6) [inline phantom.py]: [nit] any reason to assign to local `lims` variable instead of just using `self.lims` everywhere?

File: phantom.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    def make_interp(self):
+        """3D interpolation
+
+        """
+
+        xinterp = np.interp([item[0] for item in self.scatterers], self.nodeIDcoords['x'], [item[0] for item in self.dispdat])
+        yinterp = np.interp([item[1] for item in self.scatterers], self.nodeIDcoords['y'], [item[1] for item in self.dispdat])
+        zinterp = np.interp([item[2] for item in self.scatterers], self.nodeIDcoords['x'], [item[2] for item in self.dispdat])
+
+        self.sdispdat = []
```
Comment (mlp6) [inline phantom.py]: Do we want this to be a list or a numpy array?

File: phantom.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        for i in range(len(xinterp)):
+            self.sdispdat.append([xinterp[i], yinterp[i], zinterp[i]])
```
Comment (mlp6) [inline phantom.py]: `append()` is a relatively "costly" approach to growing a list; could do more directly (and less iteratively) using a numpy array.

File: phantom.py (hunk 5 above)
Comment (mlp6) [reply phantom.py]: Actually, you do import this in one of the methods below, so this global import can probably go.

**Review by mlp6 — COMMENTED** [review]
first-pass comments

File: phantom.py (hunk 5 above)
Comment (mjr52) [reply phantom.py]: changed to imported in methods

File: phantom.py (hunk 6 above)
Comment (mjr52) [reply phantom.py]: no reason, changed this

File: phantom.py (hunk 7 above)
Comment (mjr52) [reply phantom.py]: numpy array works better

File: phantom.py (hunk 8 above)
Comment (mjr52) [reply phantom.py]: fixed

File: tests/test_phantom.py (hunk 4 above)
Comment (mjr52) [reply tests/test_phantom.py]: deleted it

**Review by mjr52 — COMMENTED** [review]
I pushed 2 commits that fixed the comments from the first code review

**Pull request closed without merging**